---
title: "Preprocessing Report"
author: "Leon Wang"
date: "6 Jan, 2026"
format: html
toc: true
---

```{r}
#| label: load-packages
#| include: false

# library(shiny)
library(tidyverse)
library(data.table)
```

## Introduction

Thank you for running this Nextflow pipeline for preprocessing Oxford Nanopore antibody sequencing data.

## 1. Aligned reads

The table below provides a summary after mapping and alignment of target antibody sequences to the reference genome provided, using the [*minimap2 package*](https://lh3.github.io/minimap2/) (v2.30) and the [*samtools package*](https://www.htslib.org) (v1.22.1).

```{r}
# Read in data
#| include: false

# Get list of files
files <- list.files(
    path = "./results/1. aligned reads/stats",
    pattern = "\\.tsv$",
    full.names = TRUE
)

# Read in files
data_list <- lapply(files, fread, na.strings = c("N/A"))
# Add in barcode from file name as id col
names(data_list) <- str_replace(basename(files), "_alignment_stats.tsv", "")

# Combine datasets together and rename columns
combined_aligned_stats <- rbindlist(data_list, idcol = "sample_id", fill = TRUE)
combined_aligned_stats <- rename(combined_aligned_stats, "QC_pass" = V1, "QC_fail" = V2, "category" = V3)

# Rename and extract required rows
rows_required <- list("total", "mapped", "mapped %", "primary", "secondary", "supplementary")
combined_aligned_stats <- combined_aligned_stats %>%
    mutate(category = ifelse(category == "total (QC-passed reads + QC-failed reads)",
        "total", category
    )) %>%
    mutate(across(
        c(QC_pass, QC_fail),
        ~ ifelse(str_detect(.x, "%"),
            as.numeric(str_replace(.x, "%", "")) / 100,
            as.numeric(.x)
        )
    )) %>%
    filter(category %in% rows_required)
```

```{r}
combined_aligned_stats %>%
    distinct(sample_id) %>%
    summarise(n = n())
```

```{r}
#| label: total-reads

ggplot(data = filter(combined_aligned_stats, category == "total"), 
aes(y = sample_id, x = QC_pass)) +
    geom_col()
```

```{r}
#| label: mapped-perc

ggplot(data = filter(combined_aligned_stats, category == "mapped %"), aes(x = QC_pass * 100)) +
    geom_density() +
    xlab("Mapped (%)") +
    ylab("Density")
```

## 2. Extracted reads

The following provides detail after extracting the heavy and light sequence sequences using matchbox [*samtools package*](https://jakob-schuster.github.io/matchbox-docs/).

```{r}
# Read in data
#| include: false

# Get list of files
counts_files <- list.files(
    path = "./results/2. extracted reads/counts",
    pattern = "\\.csv$",
    full.names = TRUE
)

# Read in files
counts_list <- lapply(counts_files, fread, na.strings = c("N/A"))
# Add in barcode from file name as id col
names(counts_list) <- str_replace(basename(counts_files), "_count.csv", "")

# Combine datasets together and rename columns
combined_counts <- rbindlist(counts_list, idcol = "sample_id", fill = TRUE)
combined_counts <- rename(combined_counts, "sequence" = value)

# Divide total reads by 2 (as duplicated when using reverse complement in matchbox)
combined_counts <- combined_counts %>%
    mutate(count = ifelse(sequence == "total reads", count / 2, count))
```

```{r}
# Read in data
#| include: false
a <- combined_aligned_stats %>%
    filter(category == "primary" | category == "total")

b <- combined_counts %>% filter(sequence == "total reads")

c <- full_join(a, b)
# Why more reads after samtools (only ~ same as primary reads) vs matchbox
# Matchbox seems to only get the primary reads
```

```{r}
# Plot total reads from matchbox
combined_counts %>%
    filter(sequence == "total reads") %>%
    mutate(sample_id = fct_reorder(sample_id, count)) %>%
    ggplot(aes(y = sample_id, x = count)) +
    geom_col()

# Plot sequence counts as fill on each column

combined_counts %>%
    ggplot() +
    geom_col(
        data = combined_counts[(combined_counts[, sequence] == "total reads"), ],
        aes(y = fct_reorder(sample_id, count), x = count)
    ) +
    geom_col(
        data = combined_counts[(combined_counts[, sequence] == "rotated"), ],
        aes(y = fct_reorder(sample_id, count), x = count),
        fill = "blue"
    ) +
    geom_col(
        data = combined_counts[(combined_counts[, sequence] == "heavy"), ],
        aes(y = fct_reorder(sample_id, count), x = count),
        fill = "green"
    ) +
    geom_col(
        data = combined_counts[(combined_counts[, sequence] == "heavy + kappa" | combined_counts[, sequence] == "heavy + lambda"), ],
        aes(y = fct_reorder(sample_id, count), x = count),
        fill = "pink"
    ) +
    labs(x = "Number of sequences")
# ADD legend


# Plot proportion of sequences extracted through matchbox
combined_counts_prop <- combined_counts %>%
    group_by(sample_id) %>%
    mutate(prop = count / count[sequence == "total reads"]) %>%
    ungroup()

combined_counts_prop <- combined_counts_prop %>%
    mutate(sequence = case_when(
        sequence %in% c("heavy + kappa", "heavy + lambda") ~ "heavy + light",
        .default = sequence
    )) %>%
    group_by(sample_id, sequence) %>%
    summarise(prop = sum(prop)) %>%
    ungroup() %>%
    mutate(sequence = ordered(sequence, levels = c("total reads", "rotated", "heavy", "heavy + light"))) %>%
    group_by(sample_id) %>%
    arrange(sample_id, desc(sequence)) %>%
    mutate(diff = prop - lag(prop, default = first(prop))) %>%
    ungroup() %>%
    select(-prop) %>%
    right_join(combined_counts_prop, by = join_by(sample_id, sequence)) %>%
    mutate(diff = ifelse(is.na(diff), prop, diff))

combined_counts_prop %>%
    ggplot() +
    geom_col(
        aes(y = fct_reorder(sample_id, count), x = diff, fill = sequence),
        position = "stack"
    ) +
    labs(x = "Proportion of sequences", y = "Sample", fill = "Sequence")
# Change light sequence colors

#### Why is heavy reads less than heavy + light totals?
```

## 3. Annotated reads

The heavy and light chain reads were annotated using the [*riot package*](https://github.com/NaturalAntibody/riot_na) (v4.0.2).

```{r}
# Read in data
#| include: false

# Get list of files
annotated_files <- list.files(
    path = "./results/3. annotated reads",
    pattern = "\\.csv$",
    full.names = TRUE
)

# Read in files
annotated_list <- lapply(annotated_files, fread, na.strings = c("N/A"))
# Add in barcode from file name as id col
names(annotated_list) <- str_replace(basename(annotated_files), "_annot_.*\\.csv", "")

# Combine datasets together and rename columns
combined_annotated <- rbindlist(annotated_list, idcol = "sample_id", fill = TRUE)
```